runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -10387,6 +10387,278 @@ inlineCompareAndSwapNative(
+   dependencies->addPostCondition(loopLimitReg, TR::RealRegister::NoReg, cg);
+   dependencies->addPostCondition(limitReg, TR::RealRegister::NoReg, cg);
+   dependencies->addPostCondition(xmmMaskReg, TR::RealRegister::NoReg, cg);
+   dependencies->addPostCondition(maskReg, TR::RealRegister::NoReg, cg);
+   dependencies->addPostCondition(indexReg, TR::RealRegister::NoReg, cg);
+   dependencies->addPostCondition(xmmChunkReg, TR::RealRegister::NoReg, cg);
+   dependencies->addPostCondition(chunkReg, TR::RealRegister::NoReg, cg);
+
+   // Labels
+   TR::LabelSymbol *begLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *endLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *loopLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *residualLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *threeOrMoreBytesLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *fiveOrMoreBytesLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *nineOrMoreBytesLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *residualTestLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *returnNoNegativesLabel = NULL;
+   TR::LabelSymbol *returnHasNegativesLabel = NULL;
+   TR::LabelSymbol *returnBooleanLabel = NULL;
+   if (recognizedMethod == TR::java_lang_StringCoding_hasNegatives)
+      {
+      returnBooleanLabel = generateLabelSymbol(cg);
+      }
+   else
+      {
+      returnNoNegativesLabel = generateLabelSymbol(cg);
+      returnHasNegativesLabel = generateLabelSymbol(cg);
+      }
+   begLabel->setStartInternalControlFlow();
+   endLabel->setEndInternalControlFlow();
+
+
+   generateLabelInstruction(TR::InstOpCode::label, node, begLabel, cg);
+
+   // index = offset
+   generateRegRegInstruction(TR::InstOpCode::MOV4RegReg, node, indexReg, offsetReg, cg);
+
+   // limit = offset + length
+   generateRegMemInstruction(TR::InstOpCode::LEA4RegMem, node, limitReg, generateX86MemoryReference(offsetReg, lengthReg, 0, 0, cg), cg);
+
+   // loopLimit = (length & -16) + offset
+   generateRegRegInstruction(TR::InstOpCode::MOV4RegReg, node, loopLimitReg, lengthReg, cg);
+   generateRegImmInstruction(TR::InstOpCode::AND4RegImm4, node, loopLimitReg, -16, cg);
+   generateRegRegInstruction(TR::InstOpCode::ADD4RegReg, node, loopLimitReg, offsetReg, cg);
+
+
+   generateLabelInstruction(TR::InstOpCode::label, node, loopLabel, cg);
+
+   // if index >= loopLimit, jump to handling the residual bytes
+   generateRegRegInstruction(TR::InstOpCode::CMP4RegReg, node, indexReg, loopLimitReg, cg);
+   generateLabelInstruction(TR::InstOpCode::JGE4, node, residualLabel, cg);
+
+   // Load 16 bytes from address [buf + index]
+   generateRegMemInstruction(TR::InstOpCode::MOVDQURegMem, node, xmmChunkReg, generateX86MemoryReference(bufReg, indexReg, 0, TR::Compiler->om.contiguousArrayHeaderSizeInBytes(), cg), cg);
+
+   // Extract bitmask of sign bits
+   generateRegRegInstruction(TR::InstOpCode::PMOVMSKB4RegReg, node, maskReg, xmmMaskReg, cg);
```

Comment:
I think this should be using `xmmChunkReg` instead.  Looking at the code that results during instruction selection, we have this:

```
[0x78f725164560]	movdqu	VRF_0088, xmmword ptr [&GPR_0080+1*GPR_0087+0x8]		# MOVDQURegReg, SymRef [#440 +8]
 [0x78f7251645f0]	pmovmskb	GPR_0086, VRF_0085		# PMOVMSKB4RegReg
 ```
 
 but after register assignment
 
 ```
  [0x78f725164560]	movdqu	xmm0, xmmword ptr [r8+1*rax+0x8]		# MOVDQURegReg, SymRef [#440 +8]
 [0x78f7251645f0]	pmovmskb	?si?, xmm0		# PMOVMSKB4RegReg
 ```
 
 So it looks like you got lucky, and it happened to assign the same vector register for both instructions.
 
 Do you know what is the significance of the question marks around the `?si?`?

Reply:
Good catch, thank you! In fact, there's no need for an `xmmMaskReg` at all anymore.

Reply:
I found code in [`TR_Debug::getName(uint32_t realRegisterIndex, TR_RegisterSizes size)`](https://github.com/eclipse-omr/omr/blob/707c74b325df68f43fcf6d720ecbdefbdfe20f41/compiler/x/codegen/X86Debug.cpp#L2123) that prints that register as `?si?` if it doesn’t know the size of the register.  I would recommend looking into how that comes to be to make sure it doesn’t represent a problem.

Reply:
At Brad's suggestion, I added the property `IA32OpProp_IntTarget` to the `PMOVMSKB4RegReg` instruction in [`omr/compiler/x/codegen/X86Ops.ins`](https://github.com/eclipse-omr/omr/blob/master/compiler/x/codegen/X86Ops.ins#L3618-L3624), and a jitlog confirms that we're now generating a register properly.
```
 [0x789e70562360]	movdqu	xmm0, xmmword ptr [r11+1*rax+0x8]		# MOVDQURegReg, SymRef [#440 +8]
 [0x789e705623f0]	pmovmskb	edi, xmm0		# PMOVMSKB4RegReg
```
The size of the output of `pmovmskb` does vary depending on the size of the input, but it doesn't get any larger than 32 bits and according to the manual it does zero extend in shorter cases, so this seems like a pretty reasonable fix. Should I open a PR against OMR where we can discuss the more broad impact of this change?